mlflow/pipelines/steps/train.py

```diff
@@ -158,6 +172,104 @@ def _run(self, output_directory):
+        exp_id = _get_experiment_id()
+
+        primary_metric_greater_is_better = self.metric_greater_is_better[self.primary_metric]
+
+        primary_metric_order = "DESC" if primary_metric_greater_is_better else "ASC"
+
+        search_max_results = 100
+        search_result = mlflow_client.search_runs(
+            experiment_ids=exp_id,
+            run_view_type=ViewType.ALL,
+            max_results=search_max_results,
+            order_by=[f"metrics.{self.primary_metric} {primary_metric_order}"],
+        )
+
+        primary_metric_name = _get_primary_metric(self.step_config)
+        custom_metric_names = [cm["name"] for cm in _get_custom_metrics(self.step_config)]
+        metric_names = self.metric_greater_is_better.keys()
+        metric_keys = [f"{metric_name}_on_data_validation" for metric_name in metric_names]
+
+        leaderboard_items = []
+        for old_run in search_result:
+            if all(metric_key in old_run.data.metrics for metric_key in metric_keys):
+                leaderboard_items.append(
+                    {
+                        "Run ID": old_run.info.run_id,
+                        "Run Time": datetime.datetime.fromtimestamp(
+                            old_run.info.start_time // 1000
+                        ),
+                        **{
+                            metric_name: old_run.data.metrics[metric_key]
+                            for metric_name, metric_key in zip(metric_names, metric_keys)
+                        },
+                    }
+                )
+
+        top_leaderboard_items = leaderboard_items[:2]
+        top_leaderboard_items = [
+            {"Model Rank": i + 1, **t} for i, t in enumerate(top_leaderboard_items)
+        ]
+
+        if (
+            len(top_leaderboard_items) == 2
+            and top_leaderboard_items[0][self.primary_metric]
+            == top_leaderboard_items[1][self.primary_metric]
+        ):
+            # If top1 and top2 model primary metrics are equal,
+            # then their rank are both 1.
+            top_leaderboard_items[1]["Model Rank"] = "1"
+
+        top_leaderboard_item_index_values = ["Best", "2nd Best"][: len(top_leaderboard_items)]
+
+        latest_model_item = {
+            "Run ID": run.info.run_id,
+            "Run Time": datetime.datetime.fromtimestamp(run.info.start_time // 1000),
+            **eval_metrics["validation"],
+        }
+
+        leaderboard_items_contains_latest_model = False
+        for i, leaderboard_item in enumerate(leaderboard_items):
+            latest_value = latest_model_item[self.primary_metric]
+            historical_value = leaderboard_item[self.primary_metric]
+            if (primary_metric_greater_is_better and latest_value >= historical_value) or (
+                not primary_metric_greater_is_better and latest_value <= historical_value
+            ):
+                leaderboard_items_contains_latest_model = True
+                break
```

Comment:
I don't understand what this section is doing, will take more time to process it later. Have to go.

Reply:
Rewrote it slightly to make things clear. It determines whether or not the latest model is on the leaderboard and computes its rank accordingly; if it's not on the leaderboard, the rank is "something greater than the length of the leaderboard"